### PR TITLE
[ws-scaler] Scaling through Ghost Workspaces

### PR DESCRIPTION
**Pull request by csweichel**
This PR introduces a new method and mode for pre-scaling workspace nodes. Prior attempts have tried to trigger whole node scale-ups and keep entire nodes warm. This PR goes a different route, it:
- introduces a new workspace type: ghost workspaces. Those are basically placeholders for actual workspaces. They make the same resource requests and pull the same images as regular workspaces, but they don't actually run any workload. Ghost workspaces are ignored by ws-manager-bridge.
- makes ws-manager delete a ghost workspace whenever it starts a regular workspace, thereby using up a placeholder
- introduces `ws-scaler` who controls, starts and renews ghost workspaces. In this component we implement our "prescaling" strategies.

### How to test
1. `watch kubectl get pods -l component=workspace,workspaceType=ghost`
2. Start a workspace on http://cw-ghost-workspaces.staging.gitpod-dev.com/workspaces/ and watch one of the ghost workspaces disappear

### Caveats
- maybe we have to explicitly wait for the ghost workspace to be deleted before we start the actual workspace. Otherwise we might not "free up" the slot before the actual workspace is to be scheduled.

### ToDo
- [x] implement `supervisor ghost` which just listens for `SIGTERM`
- [x] ~create a deployment for ws-scaler~ fold scaler into ws-scheduler
- [x] come up with a sensible strategy: see `SwitchedSetpointController`

### CODE REVIEW

**Comment by geropl** [comment]
> come up with a sensible strategy

Suggestion: Have a time-based constant offset as first strategy. Example:

```
timeBasedConstantOffsetConfig:
  ranges:
  - range: 08:00-10:00
    offset: 20
  - range: 12:00-14
    offset: 10
  - range: 14:00-22:00
    offset: 30
```
Reasoning:
 - simple enough to serve as bottom line for further controllers
 - effective enough for 90% of the usecases

**Comment by csweichel** [comment]
> > come up with a sensible strategy
> 
> Suggestion: Have a time-based constant offset as first strategy. 

💯 That's exactly what the `TimedSwitchPointController` implements :)

**Comment by geropl** [comment]
Starting review now...

File: components/ee/ws-scheduler/pkg/scaler/controller.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+		return nil
+	}
+
+	tod := time.Date(0, 1, 1, t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), t.Location())
+	for i, sp := range c.Setpoints {
+		spt := time.Time(sp.Time)
+		if tod.Equal(spt) {
+			return &sp
+		}
+		if tod.After(spt) {
```
Comment (geropl) [inline components/ee/ws-scheduler/pkg/scaler/controller.go:201]: This line makes we wonder how `c.Setpoints` is sorted. I would have assumed ascending, but for this to make sense it must be the other way around...?

File: components/ee/ws-scheduler/pkg/scaler/controller.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+			continue
+		}
+		if i == 0 {
```
Comment (geropl) [inline components/ee/ws-scheduler/pkg/scaler/controller.go:204]: And what does this mean? It always make sure we never get passed the first setpoint in the array?
Maybe I'm not seeing the logic here, in that case: please add a comment! :-D

File: components/ee/ws-scheduler/pkg/scaler/controller.go (hunk 1 above)
Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/controller.go:201]: `c.Setpoints` is sorted ascending in NewSwitchedSetpointController ... I'll add a comment.

File: components/ee/ws-scheduler/pkg/scaler/controller.go (hunk 2 above)
Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/controller.go:204]: see above :)

File: components/ee/ws-scheduler/pkg/scaler/driver.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	state := make(map[string]workspaceState)
+	produceStatus := func() workspaceStatus {
+		var res workspaceStatus
+		res.DeletionCandidates = make([]string, 0, len(state))
+		for id, s := range state {
+			switch s.Type {
+			case api.WorkspaceType_GHOST:
+				res.Count.Ghost++
+			case api.WorkspaceType_PREBUILD:
+				res.Count.Prebuild++
```
Comment (geropl) [inline components/ee/ws-scheduler/pkg/scaler/driver.go:297]: We have configurations in which prebuilds are running in a separate NodePool, and deleting them would degrade the service without any help in scheduling workspaces.

Worth considering here?

Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:297]: at the moment none of the controllers works specifically on prebuilds. This is more of a "we might need this in the future if we want to speed up prebuilds in the same way" kind of thing.

We could even remove this for now to avoid confusion.

File: components/ee/ws-scheduler/pkg/scaler/driver.go [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+		case status = <-statusChan:
+			for _, id := range status.DeletionCandidates {
+				delete(startingGhosts, id)
+			}
+			status.Count.Ghost += len(startingGhosts)
+			counts <- status.Count
+			log.WithField("counts", counts).Debug("status update")
+		case <-houseKeeping.C:
+			for id, t1 := range startingGhosts {
+				if time.Since(t1) < maxGhostStartTime {
```
Comment (geropl) [inline components/ee/ws-scheduler/pkg/scaler/driver.go:136]: Why is this necessary? Are there ghosts taking too long to start? What are possible reasons?

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 3 above)
Comment (geropl) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:297]: :+1:
Also, I don't get why `REGULAR` workspaces are added to `DeletionCandidates` as well. `DeletionCandidates` is passed to `stopGhostWorkspaces`, which in turn just stops them without any filtering for `type == GHOST` like the name suggests.

File: components/ee/ws-scheduler/pkg/scaler/driver.go [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+			if d == 0 {
+				continue
+			}
+			if wspd.Config.ReactionDelay > 0 {
+				time.Sleep(time.Duration(wspd.Config.ReactionDelay))
+			}
+			var (
+				err error
+				ids []string
+			)
```
Comment (geropl) [inline components/ee/ws-scheduler/pkg/scaler/driver.go:175]: Uff. The following passage has multiple intended "fallthroughs" which make it more compact but not straight forward to read... feels a tad to "clever" for my taste.

File: components/ee/ws-scheduler/pkg/scaler/driver.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+			for _, id := range ids {
+				startingGhosts[id] = time.Now()
+				status.Count.Ghost++
+			}
+		case d := <-delta:
+			if d == 0 {
+				continue
+			}
+			if wspd.Config.ReactionDelay > 0 {
+				time.Sleep(time.Duration(wspd.Config.ReactionDelay))
```
Comment (geropl) [inline components/ee/ws-scheduler/pkg/scaler/driver.go:170]: Could you elaborate on the reason for this? Would it make sense to refresh `status` after such a delay and before acting on it (even defensively?).

File: components/ee/ws-scheduler/pkg/scaler/controller.go (hunk 2 above)
Comment (geropl) [reply components/ee/ws-scheduler/pkg/scaler/controller.go:204]: Now it makes sense. :) I'd prefer
```
if tod.Before(time.Time(c.Setpoints[0].Time)) {
  return nil
}
```
which would make the loop "cleaner" (less special cases). But that's a matter of taste.

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 5 above)
Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:179]: Unfortunately I don't see this. I've added a couple of breaklines to try and make the structure clearer.
Did that help?

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 6 above)
Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:173]: I've added a comment here.

We could, instead of just delaying everything, implement a scheme we where we collect updates and if there hasn't been an update in the last X time, or we've reached N updates (after which we'll forcefully wait X time), run the controller. This would probably make the controller less "hectic".

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 3 above)
Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:302]: that is a 🐛 

good catch!

File: components/ee/ws-scheduler/pkg/scaler/controller.go (hunk 2 above)
Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/controller.go:206]: indeed ... I did have this in a version, but replaced it with the `i == 0` case because it explicitly guards against the `i = 0; c.Setpoints[i - 1]` case. With the the `tod.Before` clause that connection isn't as obvious.

All in all I'm not terribly happy with this function. It's too hard to understand for what it does.

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 4 above)
Comment (csweichel) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:136]: It's more of a fallback. I wanted to avoid the situation where we're accumulating ghosts that don't actually exist.
This case should not happen, as we're always listening to updates/receive a full status list when we reconnect.
I'll put a warning there, so in case it it does happen we'll learn about it.

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 5 above)
Comment (geropl) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:181]: Yes, better.

Just to explain what I meant, no need to take action:
 - `if ... if` in comparison to `if ... else`: The latter better guides the eye IMO (taste)
 - both cases can definitely result in errors. Having the fallthrough is nice because it reduces dublication but also needs another "eye-jump" to parse correctly - especially if combined with the `if ... if` from above (again, taste)

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 4 above)
Comment (geropl) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:136]: Hm. I think what I'm stumbling about is the fact that:
 - we have the controller output a `delta` instead of an absolute target value, which makes it more likely to change over time
 - we even increase the gap between calculation and execution

File: components/ee/ws-scheduler/pkg/scaler/driver.go (hunk 6 above)
Comment (geropl) [reply components/ee/ws-scheduler/pkg/scaler/driver.go:175]: I'd even argue for a approach where we'd constantly update the status but only call the controller in a fixed interval. That would make it far easier to reason about it, plus it's easier to guarantee that it is stable (not "hectic") as opposed to escalating itself.

**Review by geropl — APPROVED** [review]
With the changes to stabilize the control/regulation loop: :heavy_check_mark:
